File: core-booking-engine/packages/payments/src/provider.ts
export type PaymentStatus =
  | 'pending'
  | 'authorized'
  | 'paid'
  | 'failed'
  | 'refunded'
  | 'canceled'

export interface CreatePaymentInput {
  appointmentId: string
  businessId: string
  amountCents: number
  currency: string
}

export interface PaymentResult {
  paymentId: string
  status: PaymentStatus
  providerReference?: string
}

export interface PaymentProvider {
  createPayment(input: CreatePaymentInput): Promise<PaymentResult>
  confirmPayment(paymentId: string): Promise<PaymentResult>
  refundPayment(paymentId: string): Promise<PaymentResult>
}

// Valid state transitions — no skipping steps.
export const VALID_TRANSITIONS: Record<PaymentStatus, PaymentStatus[]> = {
  pending:    ['authorized', 'failed'],
  authorized: ['paid', 'refunded', 'failed'],
  paid:       ['refunded'],
  failed:     [],
  refunded:   [],
  canceled:   [],
}

export function assertValidTransition(current: PaymentStatus, next: PaymentStatus): void {
  if (!VALID_TRANSITIONS[current].includes(next)) {
    throw new Error(`Transición inválida: ${current} → ${next}`)
  }
}

File: backend/app/payments/provider.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Literal


@dataclass
class CreatePaymentInput:
    appointment_id: str
    business_id: str
    amount_cents: int
    currency: str


@dataclass
class PaymentResult:
    payment_id: str
    status: Literal["pending", "authorized", "paid", "failed", "refunded"]
    provider_reference: str | None = None
    client_secret: str | None = None  # Stripe only — used by frontend Elements


class PaymentProvider(ABC):
    @abstractmethod
    async def create_payment(self, input: CreatePaymentInput) -> PaymentResult: ...

    @abstractmethod
    async def confirm_payment(self, payment_id: str) -> PaymentResult: ...

    @abstractmethod
    async def refund_payment(self, payment_id: str) -> PaymentResult: ...
